File: go/src/permutations/permutations.go
// 2017-11-14 18:54
package main

import "fmt"

func dfs(record []int, visited map[int]bool, nums []int, ret *[][]int) {
	if len(nums) == len(record) {
		tmp := make([]int, len(record))
		copy(tmp, record)
		*ret = append(*ret, tmp)
		return
	}
	for i := 0; i < len(nums); i++ {
		if visited[i] {
			continue
		}
		record = append(record, nums[i])
		visited[i] = true
		dfs(record, visited, nums, ret)
		visited[i] = false
		record = record[:len(record)-1]
	}
}

func permute(nums []int) [][]int {
	ret := [][]int{}
	visited := make(map[int]bool)
	dfs([]int{}, visited, nums, &ret)
	return ret
}
func main() {
	fmt.Println(permute([]int{1, 2, 3}))
}

File: java/Permutations.java
// Accepted
import java.util.ArrayList;
import java.util.List;
// @lc id=46 lang=java
// problem: permutations
class Solution {
    private void backtrack(List<List<Integer >> ret, ArrayList<Integer> curr, int [] nums, boolean [] visited) {
        if (curr.size() == nums.length) {
            ret.add(new ArrayList(curr));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (visited[i]) {
                continue;
            }
            visited[i] = true;
            curr.add(nums[i]);
            backtrack(ret, curr, nums, visited);
            curr.remove(curr.size() - 1);
            visited[i] = false;
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer >> ret = new ArrayList();
        backtrack(ret, new ArrayList<Integer>(), nums, new boolean[nums.length]);
        return ret;
    }
}
public class Permutations {
    public static void main(String[] argv) {
        System.out.println(new Solution().permute(new int[] {1, 2, 3}));
    }
}
